File: src/main/java/no/nav/arbeidsgiver/sykefravarsstatistikk/api/statistikk/sykefraværshistorikk/oppsummert/TrendKalkulator.java
package no.nav.arbeidsgiver.sykefravarsstatistikk.api.statistikk.sykefraværshistorikk.oppsummert;

import static no.nav.arbeidsgiver.sykefravarsstatistikk.api.felles.ÅrstallOgKvartal.SISTE_PUBLISERTE_KVARTAL;
import static no.nav.arbeidsgiver.sykefravarsstatistikk.api.statistikk.sykefraværshistorikk.UmaskertSykefraværForEttKvartal.hentUtKvartal;

import io.vavr.control.Either;
import java.math.BigDecimal;
import java.util.List;
import java.util.Optional;
import lombok.AllArgsConstructor;
import lombok.EqualsAndHashCode;
import lombok.ToString;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.felles.ÅrstallOgKvartal;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.statistikk.Statistikkategori;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.statistikk.sykefraværshistorikk.ManglendeDataException;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.statistikk.sykefraværshistorikk.UmaskertSykefraværForEttKvartal;

@ToString
@EqualsAndHashCode
@AllArgsConstructor
class TrendKalkulator {

    private static final ÅrstallOgKvartal nyesteKvartal = SISTE_PUBLISERTE_KVARTAL;
    private static final ÅrstallOgKvartal ettÅrTidligere =
            SISTE_PUBLISERTE_KVARTAL.minusEttÅr();
    public BigDecimal trendverdi;
    public int antallTilfellerIBeregningen;
    public List<ÅrstallOgKvartal> kvartalerIBeregningen;

    public static Either<ManglendeDataException, TrendKalkulator> kalkulerTrend(
            List<UmaskertSykefraværForEttKvartal> sykefravær) {

        Optional<UmaskertSykefraværForEttKvartal> nyesteSykefravær =
                hentUtKvartal(sykefravær, nyesteKvartal);
        Optional<UmaskertSykefraværForEttKvartal> sykefraværetEtÅrSiden =
                hentUtKvartal(sykefravær, ettÅrTidligere);

        if (nyesteSykefravær.isEmpty() || sykefraværetEtÅrSiden.isEmpty()) {
            return Either.left(
                    new ManglendeDataException(
                            "Mangler " + nyesteKvartal + " eller " + ettÅrTidligere));
        }

        BigDecimal trendverdi = nyesteSykefravær.get().getProsent()
                .subtract(sykefraværetEtÅrSiden.get().getProsent());
        int antallTilfeller =
                nyesteSykefravær.get().getAntallPersoner()
                        + sykefraværetEtÅrSiden.get().getAntallPersoner();

        return Either.right(
                new TrendKalkulator(
                        trendverdi,
                        antallTilfeller,
                        List.of(SISTE_PUBLISERTE_KVARTAL, SISTE_PUBLISERTE_KVARTAL.minusEttÅr())
                ));
    }

    public OppsummertStatistikkDto tilOppsummertStatistikkDto(
            Statistikkategori type, String label) {
        return new OppsummertStatistikkDto(
                type,
                label,
                this.trendverdi.toString(),
                this.antallTilfellerIBeregningen,
                this.kvartalerIBeregningen);
    }
}

File: src/main/java/no/nav/arbeidsgiver/sykefravarsstatistikk/api/applikasjon/aggregering/Trendkalkulator.kt
package no.nav.arbeidsgiver.sykefravarsstatistikk.api.applikasjon.aggregering

import arrow.core.Either
import arrow.core.left
import arrow.core.right
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.applikasjon.domenemodeller.Trend
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.applikasjon.domenemodeller.UmaskertSykefraværForEttKvartal
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.applikasjon.domenemodeller.ÅrstallOgKvartal
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.applikasjon.exceptions.UtilstrekkeligDataException

data class Trendkalkulator(
    var datagrunnlag: List<UmaskertSykefraværForEttKvartal>,
    var sistePubliserteKvartal: ÅrstallOgKvartal
) {
    fun kalkulerTrend(): Either<UtilstrekkeligDataException, Trend> {
        val ettÅrSiden = sistePubliserteKvartal.minusEttÅr()
        val nyesteSykefravær = hentUtKvartal(datagrunnlag, sistePubliserteKvartal)
        val sykefraværetEttÅrSiden = hentUtKvartal(datagrunnlag, ettÅrSiden)
        if (nyesteSykefravær == null || sykefraværetEttÅrSiden == null) {
            return UtilstrekkeligDataException(
                "Mangler data for $sistePubliserteKvartal og/eller $ettÅrSiden"
            ).left()
        }
        val nyesteSykefraværsprosent = nyesteSykefravær.kalkulerSykefraværsprosent().getOrNull()
        val sykefraværsprosentEttÅrSiden = sykefraværetEttÅrSiden.kalkulerSykefraværsprosent().getOrNull()
        if (nyesteSykefraværsprosent == null || sykefraværsprosentEttÅrSiden == null) {
            return UtilstrekkeligDataException("Feil i utregningen av sykefraværsprosenten, kan ikke regne ut trendverdi.").left()
        }
        val trendverdi = nyesteSykefraværsprosent.subtract(sykefraværsprosentEttÅrSiden)
        val antallTilfeller = (nyesteSykefravær.antallPersoner
                + sykefraværetEttÅrSiden.antallPersoner)
        return Trend(trendverdi, antallTilfeller, listOf(sistePubliserteKvartal, ettÅrSiden)).right()
    }
}

fun hentUtKvartal(
    sykefravær: Collection<UmaskertSykefraværForEttKvartal>, kvartal: ÅrstallOgKvartal
): UmaskertSykefraværForEttKvartal? {
    return sykefravær.find { it.årstallOgKvartal == kvartal }
}
